src/current/advisories/a99561.md

```diff
@@ -0,0 +1,51 @@
+---
+title: Technical Advisory 99561
+advisory: A-99561
+summary: When adding a new column using the new declarative schema changer, CockroachDB retrieves a unique secondary index mistakenly, assuming it to be a primary index.
+toc: true
+affected_versions: v22.2.x, v23.1.0-v23.1.5
+advisory_date: 2023-07-10
+docs_area: releases
+---
+
+Publication date: {{ page.advisory_date | date: "%B %e, %Y" }}
+
+## Description
+
+CockroachDB will automatically create a primary [index](../v23.1/indexes.html) for each table, which indexes the table's [primary key](../v23.1/primary-key.html), or another unique value when there is no primary key. A secondary index is also created automatically for columns with a [`UNIQUE`](../v23.1/unique.html) constraint. In CockroachDB v20.2 and v21.1 (the first two versions to support [`ALTER PRIMARY KEY`](../v21.1/alter-primary-key.html)), there was a bug where in some cases the automatically created index would have its `EncodingType` field marked as `PrimaryIndexEncoding`. This behavior was fixed in [#66599](https://github.com/cockroachdb/cockroach/pull/66599), but left [secondary indexes](../v23.1/schema-design-indexes.html) with primary encoding.
+
+In CockroachDB v22.2 and later, the implementation for [`ALTER TABLE`](../v22.2/alter-table.html#add-column) began using the new [declarative schema changer](../v23.1/online-schema-changes.html#declarative-schema-changer) framework. When adding a new column using the new declarative schema changer, CockroachDB retrieves a unique secondary index mistakenly, assuming it to be a primary index. As a result, the secondary index is modified to store the new columns. CockroachDB expects there to be only one primary index, but in the case outlined here, there are two, because a secondary index has been incorrectly labeled as a primary one. This will cause the new column data to be present only in the secondary index and could lead to irrecoverable data loss, if the secondary index is deleted by the user.
+
+## Statement
+
+This is resolved in CockroachDB by [#105828](https://github.com/cockroachdb/cockroach/pull/105828), which makes CockroachDB automatically repair the encoding type for secondary indexes when accessing the table. This PR will also add validation to detect tables with this same problem so that they can be fixed.
```

Comment:
The last line can be removed as the validation is not being backported in this PR.

```suggestion
This is resolved in CockroachDB by [#105828](https://github.com/cockroachdb/cockroach/pull/105828), which makes CockroachDB automatically repair the encoding type for secondary indexes when accessing the table.
```